osu.Game/Screens/Play/SaveFailedScoreButton.cs

```diff
@@ -0,0 +1,140 @@
+        {
+            InternalChild = shakeContainer = new ShakeContainer
+            {
+                RelativeSizeAxes = Axes.Both,
+                Child = button = new DownloadButton
+                {
+                    RelativeSizeAxes = Axes.Both,
+                }
+            };
+
+            button.Action = () =>
+            {
+                switch (State.Value)
+                {
+                    case ImportState.Imported:
+                        game?.PresentScore(score, ScorePresentType.Gameplay);
+                        break;
+
+                    case ImportState.Importing:
+                        shakeContainer.Shake();
+                        break;
+
+                    default:
+                        saveScore();
+                        break;
+                }
+            };
+            State.BindValueChanged(state =>
+            {
+                switch (state.NewValue)
+                {
+                    case ImportState.Imported:
+                        button.State.Value = DownloadState.LocallyAvailable;
+                        break;
+
+                    case ImportState.Importing:
+                        button.State.Value = DownloadState.Importing;
+                        break;
+
+                    case ImportState.Failed:
+                        button.State.Value = DownloadState.NotDownloaded;
+                        break;
+                }
+            }, true);
+            State.BindValueChanged(updateTooltip, true);
+        }
+
+        private void saveScore()
+        {
+            State.Value = ImportState.Importing;
+            saveReplayAsync = Task.Run(SaveReplay);
+
+            saveScoreDelegate = new ScheduledDelegate(() =>
+            {
+                if (saveReplayAsync?.IsCompleted != true)
+                    // If the asynchronous preparation has not completed, keep repeating this delegate.
+                    return;
+
+                saveScoreDelegate?.Cancel();
+
+                score = saveReplayAsync.GetAwaiter().GetResult();
+
+                State.Value = score != null ? ImportState.Imported : ImportState.Failed;
+            }, Time.Current, 50);
+
+            Scheduler.Add(saveScoreDelegate);
+        }
+
+        private void updateTooltip(ValueChangedEvent<ImportState> state)
+        {
+            switch (state.NewValue)
+            {
+                case ImportState.Imported:
+                    button.TooltipText = @"Watch replay";
+                    break;
+
+                case ImportState.Importing:
+                    button.TooltipText = @"Importing score";
+                    break;
+
+                case ImportState.Failed:
+                    button.TooltipText = @"Import failed, click button to re-import";
+                    break;
+
+                default:
+                    button.TooltipText = @"Save score";
+                    break;
+            }
+        }
```

Comment:
I don't know about this. We've never used a tooltip to show progress of an ongoing operation.

Reply:
I borrowed from [here](https://github.com/ppy/osu/pull/10339/files#diff-82cb1b3ae319044e80814e92987d37d68782835752177cbab90f610eca2bbad5R31-R43)

https://github.com/ppy/osu/blob/master/osu.Game/Beatmaps/Drawables/Cards/Buttons/DownloadButton.cs#L55

I will rename it to `updateState()`

by the way it should be possible to do localization here, but it seems not now?

Reply:
Localisation is still a bit of an afterthought, probably fine to leave for this PR (there's enough to fix already).